### ISSUE TITLE
Contracts: change `ensure_no_start` to `start`

### ISSUE THREAD

**Issue by yjhmelody** [paritytech/substrate#13621]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Motivation

Some languages must init basic env(GC) for later call.
It's very hard to remove `start` function.

### Request

Support to call `start` before `deploy`/`call`

### Solution

Also compute gas cost for `start` function.

### Are you willing to help with this request?

Maybe (please elaborate above)

**Comment by athei** [comment]
You are right this is a breaking change. I am sorry that we broke your use case. We assumed nobody would be using these functions. I understand this is annoying. We can certainly bring it back if necessary.

I understand that "it broke my code" and "it is in the wasm spec" are all valid arguments. But I am genuinely interested  if ask! needs this function. Can't the initializer code be called from the beginning of each of the two entry points (`call` and `deploy`)? They are guaranteed to be only called once per instance.

**Comment by dcodeIO** [comment]
In AssemblyScript, the start function is compiler-generated and cannot be called from user code. It is also not guaranteed to exist, depending on whether a module has reachable top-level statements that might, if part of the standard library, initialize the memory manager / GC. Ideally, a runtime would run the denoted `(start ...)` function of a Wasm module automatically, either during instantiation, or before the first export is called.

**Comment by athei** [comment]
So essentially: It is outside of the control of ask! because the AssemblyScript compiler assumes this functionality?

**Comment by athei** [comment]
We should implement that after #13639. Then wasmi will take care of metering the `start` function.

**Comment by dcodeIO** [comment]
> So essentially: It is outside of the control of ask! because the AssemblyScript compiler assumes this functionality?

It is theoretically possible to post-process a binary, i.e. if an implicit start function is present, remove the `(start ...)` entry of the module and add a bit of boilerplate to every externally callable function. As a concrete example, an `i32` global `is_start_called` could be added, and at the beginning of every externally callable function (these are function exports, and table entries if the table is exported) check whether `!is_start_called`. If `true`, set `is_start_called = true` and call what previously was the implicit start function. Mentioning in case such a workaround turns out useful.

**Comment by yjhmelody** [comment]
Hi @athei Is there any plan for this function recently?

**Comment by athei** [comment]
Is planned to be done right after https://github.com/paritytech/substrate/issues/13639 which @agryaznov is on.
